File: backend/src/workflows/infrastructure/builders/workflow_rule.py
from src.common.database.models.processing.workflow_rule import WorkflowRuleORM
from src.common.domain.models.processing.workflow_rule import WorkflowRule


def build_workflow_rule(orm_instance: WorkflowRuleORM) -> WorkflowRule:
    return WorkflowRule(
        uuid=orm_instance.uuid,
        tenant_id=orm_instance.tenant_id,
        workflow_id=orm_instance.workflow_id,
        name=orm_instance.name,
        slug=orm_instance.slug,
        position=orm_instance.position,
        is_active=orm_instance.is_active,
        kind=orm_instance.kind,
        prompt=orm_instance.prompt,
        when=orm_instance.when_expr,
        config=orm_instance.config or {},
        scope=orm_instance.scope or {"mode": "ALL_DOCUMENTS", "on_empty": "SKIPPED"},
        knowledge_refs=list(orm_instance.knowledge_refs or []),
        current_compilation_id=orm_instance.current_compilation_id,
        created_at=orm_instance.created_at,
        updated_at=orm_instance.updated_at,
    )

File: frontend/src/domain/entities/workflow-rule.ts
/**
 * Domain entities for the redesigned workflow rules system (spec §3).
 * Names use the WorkflowRule* prefix; file is short for ergonomic imports.
 */

export type WorkflowRuleScopeMode =
  | "SINGLE_DOCUMENT"
  | "TUPLE_CARTESIAN"
  | "AGGREGATE_OVER_TYPE"
  | "ALL_DOCUMENTS";

export type WorkflowRuleOnEmpty = "SKIPPED" | "FAILED" | "PASSED";

export interface WorkflowRuleScope {
  mode: WorkflowRuleScopeMode;
  documentType?: string | null;
  documentTypes?: string[] | null;
  onEmpty: WorkflowRuleOnEmpty;
}

export interface WorkflowRule {
  uuid: string;
  tenantId: string;
  workflowId: string;
  name: string;
  position: number;
  isActive: boolean;
  kind: string;
  prompt: string;
  config: Record<string, unknown>;
  scope: WorkflowRuleScope;
  knowledgeRefs: string[];
  currentCompilationId: string | null;
  createdAt: string | null;
  updatedAt: string | null;
}

export interface CreateWorkflowRulePayload {
  name: string;
  kind: string;
  prompt: string;
  config?: Record<string, unknown>;
  scope?: WorkflowRuleScope;
  knowledgeRefs?: string[];
  isActive?: boolean;
}

export interface UpdateWorkflowRulePayload {
  name?: string;
  kind?: string;
  prompt?: string;
  config?: Record<string, unknown>;
  scope?: WorkflowRuleScope;
  knowledgeRefs?: string[];
  isActive?: boolean;
}
